File: examples/cron-agent.py
#!/usr/bin/env python3
"""
BV-7X Cron Agent (Python)

Non-interactive daily agent designed for cron scheduling.
Fetches market data, applies a multi-factor strategy, and places
a blind prediction during the 21:00-22:00 UTC window.

Exits silently if the prediction window is closed (cron-safe).

Setup:
    export BV7X_API_KEY="bv7x_your_key_here"
    export BV7X_AGENT_NAME="my-agent"  # optional, for result logging

Cron:
    # Run at 21:15 UTC daily (15 min into blind window)
    15 21 * * * BV7X_API_KEY=your_key /usr/bin/python3 /path/to/cron-agent.py >> /var/log/bv7x-cron.log 2>&1

Requires: Python 3.7+, requests (`pip install requests`)
"""

import json
import os
import sys
from datetime import datetime, timezone
from pathlib import Path

import requests

BASE_URL = "https://bv7x.ai/api/bv7x"
LOG_FILE = Path(__file__).parent / ".bv7x-cron-log.json"


def log(msg):
    ts = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S UTC")
    print(f"[{ts}] {msg}")


def load_log():
    if LOG_FILE.exists():
        try:
            return json.loads(LOG_FILE.read_text())
        except (json.JSONDecodeError, OSError):
            pass
    return {"predictions": []}


def save_log(data):
    LOG_FILE.write_text(json.dumps(data, indent=2))


def fetch_json(url):
    resp = requests.get(url, timeout=15)
    resp.raise_for_status()
    return resp.json()


def check_window():
    """Check if prediction window is open. Returns None if closed."""
    data = fetch_json(f"{BASE_URL}/arena/current-round")
    if data.get("prediction_open"):
        return data
    return None


def get_market_data():
    """Fetch signal endpoint for market context."""
    return fetch_json(f"{BASE_URL}/openclaw/signal")


def strategy(market):
    """
    Multi-factor strategy for daily BTC direction prediction.

    Factors:
    1. Fear & Greed extremes (contrarian)
    2. Oracle signal direction + confidence
    3. Signal strength
    """
    fg = market.get("fearGreed", {}).get("value")
    pars = market.get("parsimonious", {})
    oracle_dir = pars.get("direction")
    oracle_conf = pars.get("confidence", 0)
    strength = market.get("signalStrength", 0)

    score = 0  # positive = UP, negative = DOWN
    reasons = []

    # Factor 1: Fear & Greed extremes (contrarian)
    if fg is not None:
        if fg <= 20:
            score += 2
            reasons.append(f"extreme_fear({fg})")
        elif fg <= 35:
            score += 1
            reasons.append(f"fear({fg})")
        elif fg >= 80:
            score -= 2
            reasons.append(f"extreme_greed({fg})")
        elif fg >= 65:
            score -= 0.5
            reasons.append(f"greed({fg})")

    # Factor 2: Oracle direction weighted by confidence
    if oracle_dir == "UP":
        weight = oracle_conf * 2
        score += weight
        reasons.append(f"oracle_UP(conf={oracle_conf:.2f})")
    elif oracle_dir == "DOWN":
        weight = oracle_conf * 2
        score -= weight
        reasons.append(f"oracle_DOWN(conf={oracle_conf:.2f})")

    # Factor 3: Signal strength
    if abs(strength) > 0.5:
        contribution = strength * 0.5
        score += contribution
        reasons.append(f"strength({strength:.2f})")

    # Decision
    if score > 0.5:
        return "UP", score, reasons
    elif score < -0.5:
        return "DOWN", score, reasons
    else:
        return None, score, reasons


def place_prediction(api_key, direction):
    """Place a blind prediction."""
    resp = requests.post(
        f"{BASE_URL}/arena/bet",
        headers={"Authorization": f"Bearer {api_key}"},
        json={"direction": direction, "round_type": "daily"},
        timeout=15,
    )
    return resp.json()


def main():
    api_key = os.environ.get("BV7X_API_KEY")
    if not api_key:
        log("ERROR: BV7X_API_KEY not set")
        sys.exit(1)

    # Check prediction window
    round_info = check_window()
    if not round_info:
        # Silent exit — cron-safe (window is closed, nothing to do)
        sys.exit(0)

    log("Prediction window is OPEN")

    # Fetch market data
    market = get_market_data()
    btc = market.get("btcPrice", 0)
    fg = market.get("fearGreed", {}).get("value", "?")
    log(f"BTC: ${btc:,.0f}, F&G: {fg}")

    # Run strategy
    direction, score, reasons = strategy(market)

    if not direction:
        log(f"SKIP: no conviction (score={score:.2f}, factors={reasons})")
        sys.exit(0)

    log(f"PREDICTION: {direction} (score={score:.2f}, factors={reasons})")

    # Place prediction
    result = place_prediction(api_key, direction)

    if result.get("success"):
        bet = result["bet"]
        blind = "BLIND " if result.get("blind") else ""
        log(f"{blind}Bet placed: {bet['direction']}, ID: {bet['id']}")

        # Log to file
        history = load_log()
        history["predictions"].append({
            "date": datetime.now(timezone.utc).strftime("%Y-%m-%d"),
            "direction": direction,
            "score": round(score, 2),
            "reasons": reasons,
            "blind": result.get("blind", False),
            "bet_id": bet["id"],
            "btc_price": bet.get("btc_price_at_bet"),
        })
        # Keep last 90 days
        history["predictions"] = history["predictions"][-90:]
        save_log(history)
    else:
        error = result.get("error", "Unknown error")
        log(f"BET FAILED: {error}")
        sys.exit(1)


if __name__ == "__main__":
    main()

File: examples/cron-agent.js
#!/usr/bin/env node
/**
 * BV-7X Cron Agent (Node.js)
 *
 * Non-interactive daily agent designed for cron scheduling.
 * Fetches market data, applies a multi-factor strategy, and places
 * a blind prediction during the 21:00-22:00 UTC window.
 *
 * Exits silently if the prediction window is closed (cron-safe).
 *
 * Setup:
 *   export BV7X_API_KEY="bv7x_your_key_here"
 *
 * Cron:
 *   # Run at 21:15 UTC daily (15 min into blind window)
 *   15 21 * * * BV7X_API_KEY=your_key /usr/bin/node /path/to/cron-agent.js >> /var/log/bv7x-cron.log 2>&1
 *
 * Requires: Node.js 18+ (uses built-in fetch)
 */

const fs = require("fs");
const path = require("path");

const BASE_URL = "https://bv7x.ai/api/bv7x";
const LOG_FILE = path.join(__dirname, ".bv7x-cron-log.json");

function log(msg) {
  const ts = new Date().toISOString().replace("T", " ").replace(/\.\d+Z/, " UTC");
  console.log(`[${ts}] ${msg}`);
}

function loadLog() {
  try {
    return JSON.parse(fs.readFileSync(LOG_FILE, "utf8"));
  } catch {
    return { predictions: [] };
  }
}

function saveLog(data) {
  fs.writeFileSync(LOG_FILE, JSON.stringify(data, null, 2));
}

async function fetchJSON(url) {
  const resp = await fetch(url);
  if (!resp.ok) throw new Error(`HTTP ${resp.status}: ${url}`);
  return resp.json();
}

async function checkWindow() {
  const data = await fetchJSON(`${BASE_URL}/arena/current-round`);
  return data.prediction_open ? data : null;
}

async function getMarketData() {
  return fetchJSON(`${BASE_URL}/openclaw/signal`);
}

/**
 * Multi-factor strategy for daily BTC direction prediction.
 *
 * Factors:
 * 1. Fear & Greed extremes (contrarian)
 * 2. Oracle signal direction + confidence
 * 3. Signal strength
 */
function strategy(market) {
  const fg = market.fearGreed?.value;
  const pars = market.parsimonious || {};
  const oracleDir = pars.direction;
  const oracleConf = pars.confidence || 0;
  const strength = market.signalStrength || 0;

  let score = 0; // positive = UP, negative = DOWN
  const reasons = [];

  // Factor 1: Fear & Greed extremes (contrarian)
  if (fg != null) {
    if (fg <= 20) {
      score += 2;
      reasons.push(`extreme_fear(${fg})`);
    } else if (fg <= 35) {
      score += 1;
      reasons.push(`fear(${fg})`);
    } else if (fg >= 80) {
      score -= 2;
      reasons.push(`extreme_greed(${fg})`);
    } else if (fg >= 65) {
      score -= 0.5;
      reasons.push(`greed(${fg})`);
    }
  }

  // Factor 2: Oracle direction weighted by confidence
  if (oracleDir === "UP") {
    const weight = oracleConf * 2;
    score += weight;
    reasons.push(`oracle_UP(conf=${oracleConf.toFixed(2)})`);
  } else if (oracleDir === "DOWN") {
    const weight = oracleConf * 2;
    score -= weight;
    reasons.push(`oracle_DOWN(conf=${oracleConf.toFixed(2)})`);
  }

  // Factor 3: Signal strength
  if (Math.abs(strength) > 0.5) {
    const contribution = strength * 0.5;
    score += contribution;
    reasons.push(`strength(${strength.toFixed(2)})`);
  }

  // Decision
  if (score > 0.5) return { direction: "UP", score, reasons };
  if (score < -0.5) return { direction: "DOWN", score, reasons };
  return { direction: null, score, reasons };
}

async function placePrediction(apiKey, direction) {
  const resp = await fetch(`${BASE_URL}/arena/bet`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${apiKey}`,
      "Content-Type": "application/json",
    },
    body: JSON.stringify({ direction, round_type: "weekly" }),
  });
  return resp.json();
}

async function main() {
  const apiKey = process.env.BV7X_API_KEY;
  if (!apiKey) {
    log("ERROR: BV7X_API_KEY not set");
    process.exit(1);
  }

  // Check prediction window
  const roundInfo = await checkWindow();
  if (!roundInfo) {
    // Silent exit — cron-safe (window is closed, nothing to do)
    process.exit(0);
  }

  log("Prediction window is OPEN");

  // Fetch market data
  const market = await getMarketData();
  const btc = market.btcPrice || 0;
  const fg = market.fearGreed?.value ?? "?";
  log(`BTC: $${btc.toLocaleString()}, F&G: ${fg}`);

  // Run strategy
  const { direction, score, reasons } = strategy(market);

  if (!direction) {
    log(`SKIP: no conviction (score=${score.toFixed(2)}, factors=${JSON.stringify(reasons)})`);
    process.exit(0);
  }

  log(`PREDICTION: ${direction} (score=${score.toFixed(2)}, factors=${JSON.stringify(reasons)})`);

  // Place prediction
  const result = await placePrediction(apiKey, direction);

  if (result.success) {
    const { bet } = result;
    const blind = result.blind ? "BLIND " : "";
    log(`${blind}Bet placed: ${bet.direction}, ID: ${bet.id}`);

    // Log to file
    const history = loadLog();
    history.predictions.push({
      date: new Date().toISOString().split("T")[0],
      direction,
      score: Math.round(score * 100) / 100,
      reasons,
      blind: result.blind || false,
      bet_id: bet.id,
      btc_price: bet.btc_price_at_bet,
    });
    // Keep last 90 days
    history.predictions = history.predictions.slice(-90);
    saveLog(history);
  } else {
    const error = result.error || "Unknown error";
    log(`BET FAILED: ${error}`);
    process.exit(1);
  }
}

main().catch((err) => {
  log(`ERROR: ${err.message}`);
  process.exit(1);
});
